### DISCUSSION TITLE
What's the best way to set arguments in rust for List[str]

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#3189] [Questions] [answered]
Assume there is a python func:
```python
def string_func(codes: List[str] | None):
    my_rust_func(codes)
```
In rust side with pyo3, `my_rust_func` could accept `Option<vec<&str>>` or `Option<vec<String>>`. What's the difference between these two and which is better?

Thank you!!

**Comment by davidhewitt** [comment, marked as answer]
You also have `Option<&PyList>`. I would describe it like this:

- `Option<&PyList>` does no type checking on the contents, so will be fastest argument to accept but you will have to accept performance cost and fallibility later when you get items from this list and need to check if they are strings.
- `Option<Vec<&str>>` has checked all contents and placed them into a Rust `Vec`. This is a much more expensive task than just accepting `&PyList`, but you are now free to work with this data purely using Rust algorithms.
- `Option<Vec<String>>` is the same but now you've also copied all the text data into new `String` structures. This is more expensive again, but helpful if you need to keep ownership of the text data after the `#[pyfunction]` ends.

**Reply by sun-rs** [reply]
Thanks for reply! So if I need to iter this List[str] and get each of them to check if this str key is in a hashmap in rust function, the fastest argument `Option<&PyList>` maybe slow in the extract part? And also  `Option<Vec<String>>` is even slower than `Option<Vec<&str>>`?

**Reply by davidhewitt** [reply]
Yes, that sounds like a reasonable summary.
